development/extensions/tutorial_notifications.rst

```diff
@@ -0,0 +1,955 @@
+
+.. csv-table::
+    :header: "Object", "Class"
+    :delim: |
+
+    ``$auth`` | :class:`\\phpbb\\auth\\auth`
+    ``$db`` | :class:`\\phpbb\\db\\driver\\driver_interface`
+    ``$language`` | :class:`\\phpbb\\language\\language`
+    ``$user`` | :class:`\\phpbb\\user`
+    ``$phpbb_root_path`` | :class:`string` phpBB root path
+    ``$php_ext`` | :class:`string` php File extension
+    ``$user_notifications_table`` | :class:`string` User notifications table
+
+Additional services
+-------------------
+As mentioned earlier in `Registering as a Symfony service`_, we want to use two additional services.
+And because we are using the ``calls`` construct, rather than overriding the parent's ``__construct()``,
+we have to define the functions that are being called.
+You can use this construct for any registered Symfony service that you may need to inject.
+
+.. code-block:: php
+
+   /** @var \phpbb\controller\helper */
+   protected $helper;
+
+   /** @var \phpbb\user_loader */
+   protected $user_loader;
+
+   /**
+    * Set controller helper.
+    *
+    * @param \phpbb\controller\helper	$helper		Controller helper object
+    * @return void
+    */
+   public function set_helper(\phpbb\controller\helper $helper)
+   {
+       $this->helper = $helper;
+   }
+
+   /**
+    * Set user loader.
+    *
+    * @param \phpbb\user_loader		$user_loader	User loader object
+    * @return void
+    */
+   public function set_user_loader(\phpbb\user_loader $user_loader)
+   {
+       $this->user_loader = $user_loader;
+   }
+
+get_type
+--------
+This should return the service identifier as defined in `Registering as a Symfony service`_.
+
+.. code-block:: php
+
+   /**
+    * Get notification type name.
+    *
+    * @return string				The notification name as defined in services.yml
+    */
+   public function get_type()
+   {
+       return 'vendor.extension.notification.type.sample';
+   }
+
+notification_option
+-------------------
+This variable defines 2 language strings for the notification to show up in the :abbr:`UCP (User Control Panel)`.
+|br| The ``group`` is for the category under which the the notification type will show up.
+|br| The ``lang`` is for the actual notification type.
+
+It is also possible to define an ``id`` in these options.
+Usually this isn't needed for most extensions.
+This ``id`` variable is used to concatenate multiple notification types into one.
+So if you have multiple notification types that should show up as a single type in the user's preferences,
+you can set the same ``id`` on all those types.
+
+If you do not wish to display this notification in the user's preferences, you can omit this variable.
+In that case also make sure to set the is_available_ to ``false``.
+
+.. code-block:: php
+
+   /**
+    * Notification option data (for outputting to the user).
+    *
+    * @var bool|array				False if the service should use it's default data
```

Comment:
Please in all these code blocks, eliminate all the spaces so these comments don't go beyond the edge of the code block area. Just use a single space, or the bare minimum of spaces needed to line up the comments with each other.